public/codenotes.php

```diff
@@ -14,8 +14,145 @@
 
 getCodeNotes($gameID, $codeNotes);
 
-RenderContentStart('Code Notes - ' . $gameData['Title']);
+RenderContentStart('Code Notes - ' . htmlspecialchars_decode($gameData['Title']));
 ?>
+<script>
+/**
+ * Toggle the editing state of a code note row and 
+ * show/hide the elements necessary to perform an edit.
+ * 
+ * @param {Element} rowEl - The row element to toggle editing state for.
+ * @param {boolean} isEditing - Whether the row is transitioning to edit mode or not.
+ */
+function setRowEditingEnabled(rowEl, isEditing) {
+    const rowElementVisibilities = [
+        { className: 'edit-btn', showWhen: !isEditing },
+        { className: 'note-display', showWhen: !isEditing },
+        { className: 'note-edit', showWhen: isEditing },
+        { className: 'save-btn', showWhen: isEditing },
+        { className: 'cancel-btn', showWhen: isEditing },
+    ];
+
+    for (const visibility of rowElementVisibilities) {
+        const targetEl = rowEl.querySelector(`.${visibility.className}`);
+        
+        if (visibility.showWhen === true) {
+            targetEl.classList.remove('hidden');
+        } else {
+            targetEl.classList.add('hidden');
+        }
+    }
+}
+
+/**
+ * Enable edit mode for a specific row.
+ * 
+ * @param {number} rowIndex - The index of the row to enable edit mode for.
+ */
+function beginEditMode(rowIndex) {
+    const rowEl = document.getElementById(`row-${rowIndex}`);
+    setRowEditingEnabled(rowEl, true);
+
+    const noteEditEl = rowEl.querySelector('.note-edit');
+
+    // Calculate the number of rows based on the number of newline characters.
+    const rowCount = (noteEditEl.value.match(/\n/g) || []).length + 1;
+
+    // Set the textarea rows attribute, ensuring a minimum of 2 rows.
+    noteEditEl.rows = Math.max(rowCount, 2);
+}
+
+/**
+ * Cancel edit mode for a specific row.
+ * 
+ * @param {number} rowIndex - The index of the row to cancel edit mode for.
+ */
+function cancelEditMode(rowIndex) {
+    const rowEl = document.getElementById(`row-${rowIndex}`);
+    setRowEditingEnabled(rowEl, false);
```

Comment:
Cancelling doesn't revert changes.

1) Edit
2) Type in "foo"
3) Cancel

Note: "foo" disappears

4) Edit

Note "foo" still present

Reply:
Resolved in https://github.com/RetroAchievements/RAWeb/pull/1440/commits/4d190e1c894739d66e10e28aa54a33ca63822634.